qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,420 @@
+            bb, lt_gate, nu=nu, lte_mu_mp1=lte_nu_mp1
+        )
+        mu, nu, lte_mu_nu = add_from_bloq_register_flat_qubits(
+            bb, lte_gate, mu=mu, nu=nu, lte_mu_nu=lte_mu_nu
+        )
+        mu, gt_mu_n = bb.add(
+            GreaterThanConstant(num_bits_mu, self.num_spin_orb // 2), x=mu, target=gt_mu_n
+        )
+        junk = bb.add(XGate(), q=junk)
+        bb.free(bb.join(np.array([lte_mu_nu, lte_nu_mp1, gt_mu_n, junk, amp])))
+        out_regs = {'mu': mu, 'nu': nu, 'succ': succ, 'eq_nu_mp1': eq_nu_mp1}
+        return out_regs
+
+
+@frozen
+class PrepareTHC(Bloq):
+    r"""State Preparation for THC Hamilontian.
+
+    Prepares the state
+
+    $$
+        \frac{1}{\sqrt{\lambda}}|+\rangle|+\rangle\left[
+            \sum_\ell^{N/2} \sqrt{t_\ell}|\ell\rangle|M+1\rangle
+            + \frac{1}{\sqrt{2}} \sum_{\mu\le\nu}^M \sqrt{\zeta_{\mu\nu}} |\mu\rangle|\nu\rangle
+        \right].
+    $$
+
+    Note we use UniformSuperpositionTHC as a subroutine as part of this bloq in
+    contrast to the reference which keeps them separate.
+
+    Args:
+        num_mu: THC auxiliary index dimension $M$
+        num_spin_orb: number of spin orbitals $N$
+        alt_mu: Alternate values for mu indices.
+        alt_nu: Alternate values for nu indices.
+        alt_theta: Alternate values for theta indices.
+        theta: Signs of lcu coefficients.
+        keep: keep values.
+        keep_bitsize: number of bits for keep register for coherent alias sampling.
+
+    Registers:
+     - mu: $\mu$ register.
+     - nu: $\nu$ register.
+     - theta: sign register.
+     - succ: success flag qubit from uniform state preparation
+     - eq_nu_mp1: flag for if $nu = M+1$
+     - plus_a / plus_b: plus state for controlled swaps on spins.
+
+    References:
+        [Even more efficient quantum computations of chemistry through
+            tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf) Fig. 2 and Fig. 3.
+    """
+
+    num_mu: int
+    num_spin_orb: int
+    alt_mu: Tuple[int, ...]
+    alt_nu: Tuple[int, ...]
+    alt_theta: Tuple[int, ...]
+    theta: Tuple[int, ...]
+    keep: Tuple[int, ...]
+    keep_bitsize: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("mu", bitsize=(self.num_mu).bit_length()),
+                Register("nu", bitsize=(self.num_mu).bit_length()),
+                Register("theta", bitsize=1),
+                Register("succ", bitsize=1),
+                Register("eq_nu_mp1", bitsize=1),
+                Register("plus_a", bitsize=1),
+                Register("plus_b", bitsize=1),
+            ]
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"PrepareTHC(num_mu={self.num_mu}, num_spin_orb={self.num_spin_orb},"
+            "keep_bitsize={self.keep_bitsize})"
+        )
+
+    @classmethod
+    def build(cls, t_l, zeta, probability_epsilon=1e-8) -> 'PrepareTHC':
```

Comment:
I suggest a more descriptive name like `from_hamiltonian[_coeffs]`.

Please type-annotate and document these args.

Reply:
done.